virtual_rainforest/core/registry.py

```diff
@@ -0,0 +1,222 @@
+        excep = RuntimeError(f"More than one model defined in in {module_name}")
+        LOGGER.critical(excep)
+        raise excep
+
+    # Trap models that do not follow the requirement that the BaseModel.model_name
+    # attribute matches the virtual_rainforest.models.model_name
+    # TODO - can we retire the model_name attribute if it just duplicates the module
+    #        name
+    model_name, model_obj = model[0]
+    if module_name_short != model_obj.model_name:
+        excep = RuntimeError(
+            f"The model_name attribute does not match the module name in {module_name}"
+        )
+        LOGGER.critical(excep)
+        raise excep
+
+    # Register the resulting single model class
+    MODULE_REGISTRY[module_name_short]["model"] = model_obj
+    LOGGER.info(f"Registered class for {module_name_short} model: {model_name}")
+
+
+def register_schema(module_name_short: str, schema_file_path: Path) -> None:
+    """Simple function to add configuration schema to the registry.
+
+    Args:
+        module_name: The name to register the schema under
+        schema_file_path: The file path to the JSON Schema file for the model
+
+    Raises:
+        ValueError: If the module name has already been used to register a schema
+    """
+
+    if (
+        module_name_short in MODULE_REGISTRY
+        and "schema" in MODULE_REGISTRY[module_name_short]
+    ):
+        excep = ValueError(
+            f"The module schema for {module_name_short} is already registered"
+        )
+        LOGGER.critical(excep)
+        raise excep
+
+    try:
+        MODULE_REGISTRY[module_name_short]["schema"] = load_schema(
+            module_name_short, schema_file_path
+        )
+    except Exception as excep:
+        LOGGER.critical(
+            f"Schema registration for {module_name_short} failed: check log"
+        )
+        raise excep
+
+    LOGGER.info(
+        "Schema registered for module %s: %s ", module_name_short, schema_file_path
+    )
+
+
+def register_constants_classes(module: ModuleType, module_name_short: str) -> None:
+    """Simple function to add a constants class to the registry.
+
+    Args:
+        module: A module object that might contain constants classes.
+        module_name_short: The short name of the module
+
+    Raises:
+        ValueError: If the model and class name have already been used to register a
+            constants class
+    """
+
+    # Create the dictionary for constants for this module
+    if "constants" not in MODULE_REGISTRY[module_name_short]:
+        MODULE_REGISTRY[module_name_short]["constants"] = {}
+
+    # Try and get the constants submodule and simply leave if none is present
+    constants_submodule = getattr(module, "constants", None)
+
+    if constants_submodule is None:
+        return
+
+    # Otherwise get any constant dataclasses and register them
+    constants_classes = (
+        class_ for class_ in getmembers(constants_submodule) if is_dataclass(class_[1])
+    )
```

Comment:
What if a module have dataclasses that have nothing to do with constants in the VirtualRainforest sense?

Reply:
That's a good point - the intention is that the `xyz.constants` submodules should only ever contain 'constants' dataclasses. However we could create a subclass of `dataclass` that makes this explicit?

Reply:
I think that would be the way to go. Forbidding users to have other dataclasses in their module to do other stuff so that VR works seems rather arbitrary.

Reply:
I can't see any reason why you couldn't just put your dataclass in another submodule, but having an explicit ConstantsDataclass base class seems clearer and we can give it shared methods.

Reply:
Filtering using generic Python objects, which is what you are doing in essence, is bound to create conflicts as chances are you will pick things that you don't want. It's like forbidding to use integers in a module because you are using integers for some very specific application of your tool. If you have a custom class and filter by that class, everyone is happier and has more freedom to do whatever they want.